### DISCUSSION TITLE
How do I get the graph on untrained data?

### DISCUSSION THREAD

**Discussion by lnajman** [lmcinnes/umap#615] [Q&A]
I want to use UMap jointly with a classifier that needs a graph as input. It works well on training data, as it is straightforward to get the graph from the mapper. However, for test data, I only get the embedding of the points, not the graph. I have seen that you can select transform_mode = 'graph' to get a graph, but I do not understand what is the output graph on the test data, and how is it related to the graph computed on the train data. 
I found the transform_mode by reading the code, but I can not find a documentation for that. It is currently not explained in the doc.
Is there a documentation that helps me to understand both what UMAP is exactly doing for generalization, and also how to get the graph it is already computing at generalization, so that I do not have to compute this graph by myself?
Ideally, I would like to obtain a graph that contains both training and testing data, but I am not sure if this is what the current code is doing.

**Comment by lmcinnes** [comment]
The current approach builds a graph where vertices are training samples, and that can be extracted as the ``graph_`` attribute. The transform method with ``transform_mode = "graph"`` returns a sparse matrix in the same format, but just for the transformed data -- so each column corresponds to a training sample, and each row corresponds to a sample that was passed to transform. Each entry is the weight of the edge between the test sample and training sample corresponding to the rows and columns for the entry.

The important point here is the contrast between train and transformed samples: train samples can have edges joining them, but transform samples only have edges to training samples. If you want to get an adjacency matrix for the combined train and test data then you can effectively do:

```python
train_graph = mapper.graph_
test_graph = mapper.transform(test_data)
full_adjacency = scipy.sparse.vstack(
    [
        scipy.sparse.hstack([train_graph, test_graph.transpose()]),
        scipy.sparse.hstack([test_graph, scipy.sparse.csr_matrix((test_graph.shape[0], test_graph.shape[0]))
    ]
)
```

All we are doing here is putting together all the blocks. If the train graph matrix is A, and the test graph matrix is B then we are simply constructing the block-wise matrix

```
--------------------
|           |       |
|           |       |
|     A     |   B'  |
|           |       |
|           |       |
---------------------
|           |       |
|     B     |   0   |
|           |       |
---------------------
```

where 0 is simply the all zero matrix (since there are no edges among the test samples).

**Reply by lnajman** [reply]
Thanks for the quick and very clear answer. 
I have just a small questions: for a given test/transform sample, how many edges to trained samples do you compute?
Is this a fixed number, maybe just the closest one?

**Reply by lmcinnes** [reply]
It should be exactly ``n_neighbors`` many edges from each test sample.

**Comment by vishnu1729** [comment]
can we only get a graph for the test data if we use ```transform_mode = graph``` while fitting the model? I need the embeddings for train and test as well as the graph for the test data. Kindly let me know how i can get both these.

**Comment by lmcinnes** [comment]
Unfortunately yes, that's the case right now.

Sorry.

**Reply by vishnu1729** [reply]
thanks! so i think i wll have to two separate UMAP models: one for graphs and one for embeddings. Appreciate the inputs.
